### PR TITLE
on Windows, strip the '\\?\' prefix from the canonical path

### PR DESCRIPTION
**Pull request by RalfJung**
This should help with https://github.com/rust-lang/rust/pull/74146#issuecomment-655723028

### CODE REVIEW

File: cargo-miri/bin.rs [hunk 1]
```diff
@@ -258,8 +258,20 @@ fn setup(subcommand: MiriCommand) {
     // Determine where the rust sources are located.  `XARGO_RUST_SRC` env var trumps everything.
     let rust_src = match std::env::var_os("XARGO_RUST_SRC") {
         Some(val) => {
-            let val = PathBuf::from(val);
-            val.canonicalize().unwrap_or(val)
+            let path = PathBuf::from(val);
+            let path = path.canonicalize().unwrap_or(path);
+
+            // On Windows, this produces a path starting with `\\?\`, which xargo cannot deal with.
+            // Strip that prefix; the resulting path should still be valid.
```
Comment (bjorn3) [inline cargo-miri/bin.rs:265]: If the path is more than 260 characters long, the `\\?\` is required See the "Maximum Path Length Limitation" section of https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file

Comment (RalfJung) [reply cargo-miri/bin.rs:265]: Oh man, Windows is hard to handle.^^
I think this should be a good enough hack to land the Miri fix in rustc now. It anyway only affetcs setups that set `XARGO_RUST_SRC` which should be rare.

Once Miri is green in rustc, we don't have to rush any more. Then I'll release a new xargo and remove the hack here in Miri. @bjorn3 do you have a Windows system you could test this on?

Comment (bjorn3) [reply cargo-miri/bin.rs:265]: No I have switched to Linux a long time ago. Wine may help with testing though.

Comment (matklad) [reply cargo-miri/bin.rs:265]: Drive by comment, but it seems to me that we only use canonicalize here to go from relative path to an absolute one. If that is the case, I suggest using `std::env::current_dir().unwrap().join(val)` as a better alternative. Note that join auto-magically handles abs path in a convenient way (= is susceptible to the directory traversal attack :0) )

Comment (RalfJung) [reply cargo-miri/bin.rs:265]: @matklad that's a great idea, I like it. :)

Comment (retep998) [reply cargo-miri/bin.rs:265]: Note that `std::env::current_dir().unwrap().join(val)` is also *incorrect* on Windows, as drive relative paths are a thing (aka `C:foo` is relative to the current directory *on that drive*).

Comment (RalfJung) [reply cargo-miri/bin.rs:265]: > aka C:foo is relative to the current directory on that drive

I don't even know what that means, and I am not sure I want to.^^

**Pull request merged**